File: src/gearbox_sim/core/constraints.py
"""Kinematic constraints for drivetrain components.

These constraints define relationships between port speeds and torques
within a single component. The drivetrain compiler uses these to
reduce the number of degrees of freedom and build the equations of motion.
"""

from abc import ABC, abstractmethod
from dataclasses import dataclass
from typing import Dict, Tuple


class KinematicConstraint(ABC):
    """Base class for kinematic constraints between ports.

    Kinematic constraints relate the speeds and/or torques at different
    ports of a component. For example, a gear ratio constraint relates
    input and output speeds by a fixed ratio.
    """

    @abstractmethod
    def get_speed_relation(self) -> Dict[str, float]:
        """Get the speed relationship coefficients.

        Returns a dict mapping port names to coefficients such that:
        sum(coeff_i * omega_i) = 0

        For example, a gear with ratio K has:
        omega_out - omega_in / K = 0
        Returns: {"input": -1/K, "output": 1}
        """
        pass

    @abstractmethod
    def get_dependent_port(self) -> str:
        """Get the name of the port whose speed is determined by others.

        This port's DOF will be eliminated during constraint reduction.
        """
        pass

    @abstractmethod
    def get_independent_ports(self) -> list[str]:
        """Get the names of ports that remain as independent DOFs."""
        pass


@dataclass
class GearRatioConstraint(KinematicConstraint):
    """Simple gear ratio constraint: omega_out = omega_in / ratio.

    This represents a fixed-ratio gear connection. The output speed
    is the input speed divided by the ratio. Torque is multiplied
    by ratio times efficiency.

    Attributes:
        input_port: Name of the input (high-speed) port
        output_port: Name of the output (low-speed) port
        ratio: Speed ratio K = omega_in / omega_out (>1 means reduction)
        efficiency: Power transmission efficiency (0-1)
    """

    input_port: str
    output_port: str
    ratio: float
    efficiency: float = 1.0

    def get_speed_relation(self) -> Dict[str, float]:
        # omega_out = omega_in / ratio
        # omega_out - omega_in / ratio = 0
        # Rearranging: -omega_in / ratio + omega_out = 0
        return {self.input_port: -1.0 / self.ratio, self.output_port: 1.0}

    def get_dependent_port(self) -> str:
        return self.output_port

    def get_independent_ports(self) -> list[str]:
        return [self.input_port]

    def transform_speed(self, omega_in: float) -> float:
        """Calculate output speed from input speed."""
        return omega_in / self.ratio

    def transform_torque(self, T_out: float) -> float:
        """Calculate input torque required for given output torque."""
        return T_out / (self.ratio * self.efficiency)

    def get_reflected_inertia(self, J_output: float) -> float:
        """Calculate inertia reflected to the input side."""
        return J_output / (self.ratio**2)


@dataclass
class WillisConstraint(KinematicConstraint):
    """Planetary gear Willis equation constraint.

    The Willis equation for a simple planetary gear set:
    omega_sun = (1 + rho) * omega_carrier - rho * omega_ring

    where rho = Z_ring / Z_sun (the planetary ratio).

    The torque balance is:
    T_sun : T_carrier : T_ring = 1 : -(1 + rho) : rho

    Attributes:
        sun_port: Name of the sun gear port (typically MG1)
        carrier_port: Name of the carrier port (typically engine)
        ring_port: Name of the ring gear port (typically output to MG2/gearbox)
        rho: Planetary ratio Z_ring / Z_sun
    """

    sun_port: str
    carrier_port: str
    ring_port: str
    rho: float

    def get_speed_relation(self) -> Dict[str, float]:
        # omega_sun = (1 + rho) * omega_carrier - rho * omega_ring
        # Rearranging: omega_sun - (1 + rho) * omega_carrier + rho * omega_ring = 0
        return {
            self.sun_port: 1.0,
            self.carrier_port: -(1.0 + self.rho),
            self.ring_port: self.rho,
        }

    def get_dependent_port(self) -> str:
        # Sun is typically eliminated (its speed is constrained by carrier and ring)
        return self.sun_port

    def get_independent_ports(self) -> list[str]:
        return [self.carrier_port, self.ring_port]

    def calc_sun_speed(self, omega_carrier: float, omega_ring: float) -> float:
        """Calculate sun gear speed from carrier and ring speeds."""
        return (1.0 + self.rho) * omega_carrier - self.rho * omega_ring

    def calc_carrier_speed(self, omega_sun: float, omega_ring: float) -> float:
        """Calculate carrier speed from sun and ring speeds."""
        return (omega_sun + self.rho * omega_ring) / (1.0 + self.rho)

    def calc_ring_speed(self, omega_carrier: float, omega_sun: float) -> float:
        """Calculate ring speed from carrier and sun speeds."""
        return ((1.0 + self.rho) * omega_carrier - omega_sun) / self.rho

    def get_torque_ratios(self) -> Tuple[float, float, float]:
        """Get the torque balance ratios (sun : carrier : ring).

        For a planetary gear in static equilibrium:
        T_sun + T_carrier + T_ring = 0 (torque balance)
        T_sun : T_carrier : T_ring = 1 : -(1+rho) : rho

        Returns:
            Tuple of (sun_ratio, carrier_ratio, ring_ratio)
        """
        return (1.0, -(1.0 + self.rho), self.rho)

    def get_inertia_coefficients(self, J_sun: float) -> Tuple[float, float, float]:
        """Get inertia coupling coefficients for 2-DOF reduction.

        When reducing from 3-DOF (sun, carrier, ring) to 2-DOF (carrier, ring)
        using the Willis constraint, the inertia matrix becomes:

        J = [J_carrier + (1+rho)^2 * J_sun,    -(1+rho) * rho * J_sun    ]
            [-(1+rho) * rho * J_sun,            J_ring + rho^2 * J_sun    ]

        Returns:
            Tuple of (J_eq_carrier_carrier, J_coupling, J_eq_ring_ring)
            where J_coupling is the off-diagonal term.
        """
        J_cc = (1.0 + self.rho) ** 2 * J_sun  # Added to carrier inertia
        J_cr = -(1.0 + self.rho) * self.rho * J_sun  # Off-diagonal coupling
        J_rr = self.rho**2 * J_sun  # Added to ring inertia
        return (J_cc, J_cr, J_rr)


@dataclass
class RigidConnectionConstraint(KinematicConstraint):
    """Rigid connection: two ports have the same speed.

    This represents a rigid coupling between two shafts or a
    parallel connection where torques sum and speeds are equal.

    Attributes:
        port_a: Name of the first port
        port_b: Name of the second port
    """

    port_a: str
    port_b: str

    def get_speed_relation(self) -> Dict[str, float]:
        # omega_a = omega_b
        # omega_a - omega_b = 0
        return {self.port_a: 1.0, self.port_b: -1.0}

    def get_dependent_port(self) -> str:
        return self.port_b

    def get_independent_ports(self) -> list[str]:
        return [self.port_a]

File: web/packages/drivetrain-sim/src/core/constraints.ts
/**
 * Kinematic constraints that relate speeds and torques between ports.
 */

/**
 * Speed relation: sum(coeff_i * omega_i) = 0
 * Maps port names to coefficients.
 */
export type SpeedRelation = Record<string, number>;

/**
 * Base interface for kinematic constraints.
 */
export interface KinematicConstraint {
  /** Get the speed relation coefficients. */
  getSpeedRelation(): SpeedRelation;

  /** Get the port whose DOF is eliminated by this constraint. */
  getDependentPort(): string;

  /** Get the ports that remain independent. */
  getIndependentPorts(): string[];
}

/**
 * Gear ratio constraint: omega_out = omega_in / ratio
 *
 * Speed relation: omega_in - ratio * omega_out = 0
 * Torque relation: T_out = T_in * ratio * efficiency
 */
export interface GearRatioConstraintConfig {
  inputPort: string;
  outputPort: string;
  ratio: number;
  efficiency?: number;
}

export class GearRatioConstraint implements KinematicConstraint {
  readonly inputPort: string;
  readonly outputPort: string;
  readonly ratio: number;
  readonly efficiency: number;

  constructor(config: GearRatioConstraintConfig) {
    this.inputPort = config.inputPort;
    this.outputPort = config.outputPort;
    this.ratio = config.ratio;
    this.efficiency = config.efficiency ?? 1.0;
  }

  getSpeedRelation(): SpeedRelation {
    // omega_in - ratio * omega_out = 0
    return {
      [this.inputPort]: 1.0,
      [this.outputPort]: -this.ratio,
    };
  }

  getDependentPort(): string {
    return this.outputPort;
  }

  getIndependentPorts(): string[] {
    return [this.inputPort];
  }

  /** Transform input speed to output speed. */
  transformSpeed(omegaIn: number): number {
    return omegaIn / this.ratio;
  }

  /** Transform output torque to input torque (back-propagation). */
  transformTorque(torqueOut: number): number {
    return torqueOut / (this.ratio * this.efficiency);
  }

  /** Get inertia reflected through the gear ratio. */
  getReflectedInertia(jOutput: number): number {
    return jOutput / (this.ratio * this.ratio);
  }
}

/**
 * Willis constraint for planetary gear sets.
 *
 * omega_sun = (1 + rho) * omega_carrier - rho * omega_ring
 *
 * where rho = Z_ring / Z_sun (typically 2-4)
 *
 * Torque balance: T_sun : T_carrier : T_ring = 1 : -(1+rho) : rho
 */
export interface WillisConstraintConfig {
  sunPort: string;
  carrierPort: string;
  ringPort: string;
  rho: number;
}

export class WillisConstraint implements KinematicConstraint {
  readonly sunPort: string;
  readonly carrierPort: string;
  readonly ringPort: string;
  readonly rho: number;

  constructor(config: WillisConstraintConfig) {
    this.sunPort = config.sunPort;
    this.carrierPort = config.carrierPort;
    this.ringPort = config.ringPort;
    this.rho = config.rho;
  }

  getSpeedRelation(): SpeedRelation {
    // omega_sun - (1+rho)*omega_carrier + rho*omega_ring = 0
    return {
      [this.sunPort]: 1.0,
      [this.carrierPort]: -(1 + this.rho),
      [this.ringPort]: this.rho,
    };
  }

  getDependentPort(): string {
    // Sun is typically eliminated (connected to MG1)
    return this.sunPort;
  }

  getIndependentPorts(): string[] {
    return [this.carrierPort, this.ringPort];
  }

  /** Calculate sun speed from carrier and ring speeds. */
  calcSunSpeed(omegaCarrier: number, omegaRing: number): number {
    return (1 + this.rho) * omegaCarrier - this.rho * omegaRing;
  }

  /** Calculate carrier speed from sun and ring speeds. */
  calcCarrierSpeed(omegaSun: number, omegaRing: number): number {
    return (omegaSun + this.rho * omegaRing) / (1 + this.rho);
  }

  /** Calculate ring speed from carrier and sun speeds. */
  calcRingSpeed(omegaCarrier: number, omegaSun: number): number {
    return ((1 + this.rho) * omegaCarrier - omegaSun) / this.rho;
  }

  /**
   * Get torque ratios: [sun, carrier, ring]
   *
   * T_sun : T_carrier : T_ring = 1 : -(1+rho) : rho
   *
   * If sun torque is T_s, then:
   * - Carrier torque = -(1+rho) * T_s
   * - Ring torque = rho * T_s
   */
  getTorqueRatios(): [number, number, number] {
    return [1.0, -(1 + this.rho), this.rho];
  }

  /**
   * Get inertia coupling coefficients for 2-DOF system (carrier, ring).
   *
   * With sun eliminated, the coupled inertia matrix from sun's inertia J_sun is:
   * J_cc = (1+rho)^2 * J_sun
   * J_cr = -(1+rho) * rho * J_sun  (off-diagonal)
   * J_rr = rho^2 * J_sun
   */
  getInertiaCoefficients(): {
    carrierCarrier: number;
    carrierRing: number;
    ringRing: number;
  } {
    const onePlusRho = 1 + this.rho;
    return {
      carrierCarrier: onePlusRho * onePlusRho,
      carrierRing: -onePlusRho * this.rho,
      ringRing: this.rho * this.rho,
    };
  }
}

/**
 * Rigid connection constraint: omega_a = omega_b
 *
 * Used for shaft couplings and parallel connections.
 */
export class RigidConnectionConstraint implements KinematicConstraint {
  readonly portA: string;
  readonly portB: string;

  constructor(portA: string, portB: string) {
    this.portA = portA;
    this.portB = portB;
  }

  getSpeedRelation(): SpeedRelation {
    // omega_a - omega_b = 0
    return {
      [this.portA]: 1.0,
      [this.portB]: -1.0,
    };
  }

  getDependentPort(): string {
    return this.portB;
  }

  getIndependentPorts(): string[] {
    return [this.portA];
  }
}
